docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,678 @@
+ *             - SGX_EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL Any parameter passed in is not valid.
+ * @retval errno Error as reported by dependent syscalls, e.g., mmap().
+ */
+int sgx_mmap_ocall(void *addr, size_t length, int prot, int flags);
+
+/*
+ * Call OS to change permissions, type, or notify EACCEPT done after TRIM
+ *
+ * @param[in] addr start address of the memory to change protections.
+ * @param[in] length length of the area.  This must be a multiple of the page size.
+ * @param[in] flags this must be OR'ed of following:
+ *            SGX_EMA_PROT_READ
+ *            SGX_EMA_PROT_WRITE
+ *            SGX_EMA_PROT_EXEC
+ *            SGX_EMA_PAGE_TYPE_TRIM: change the page type to PT_TRIM, implies RW.
+ *            SGX_EMA_PAGE_TYPE_TCS: change the page type to PT_TCS
+ *            SGX_EMA_PROT_NOACCESS: Signal the kernel EACCEPT is done for PT_TRIM pages.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL if any parameter passed in is not valid.
+ * @retval errno Error as reported by dependent syscalls, e.g., mprotect().
+ */
+
+int sgx_mprotect_ocall(void *addr, size_t length, int flags);
+
+/*
+ * OCall to make regular munmap syscall.
+ * @param[in] addr start address of the memory to unmap.
+ * @param[in] length length of the area.  This must be a multiple of the page size.
+ * @retval 0 The operation was successful.
+ * @retval errno Error as reported by munmap syscall.
+ */
+int sgx_munmap_ocall(void *addr, size_t length);
+
+
+```
+
+### Other Utilities
+
+```
+/*
+ * Define a mutex and init/lock/unlock/destroy functions.
+ */
+typedef struct sgx_mutex_t;
+int sgx_mutex_init(sgx_mutex_t* mutex);
+int sgx_mutex_lock(sgx_mutex_t* mutex);
+int sgx_mutex_unlock(sgx_mutex_t* mutex);
+int sgx_mutex_destroy(sgx_mutex_t* mutex);
+
+/*
+ * Check whether the given buffer is strictly within the enclave.
+ *
+ * Check whether the buffer given by the **ptr** and **size** parameters is
+ * strictly within the enclave's memory. If so, return true. If any
+ * portion of the buffer lies outside the enclave's memory, return false.
+ *
+ * @param[in] ptr The pointer pointer to buffer.
+ * @param[in] size The size of the buffer
+ *
+ * @retval true The buffer is strictly within the enclave.
+ * @retval false At least some part of the buffer is outside the enclave, or
+ * the arguments are invalid. For example, if **ptr** is null or **size**
+ * causes arithmetic operations to wrap.
+ *
+ */
+bool sgx_is_within_enclave(const void *ptr, size_t size);
+
+```
+
+### Support for EMM Initialization
+
+In addition to implement the abstraction layer APIs, a runtime shall provide
+iniitial enclave memory layout information to the EMM during early
+initialization phase of the enclave.
+The memory manager must be initialized in first ECALL (ECMD_INIT_ENCLAVE in
+Intel SGX SDK) before any other clients can use it. Therefore, code and data
```

Comment:
When this document says "Intel SGX SDK" does it refer to both the Intel SGX SDK for Linux and the Intel SGX SDK for Windows, since those are different SDKs?  Clarify.  Avoid ever using the term "Intel SGX SDK" without "for ..." since the term is ambiguous/confusing.